In [95]:
import numpy as np
import pandas as pd

In [96]:
def printM3(matrix3, text):
    print(f"\n{text}:\n[")
    for i in range(len(matrix3)):
        print(matrix3[i])
    print("]")

In [97]:
def calcFoward(values, weights):
    #print(f"values:{values}\nweights:{weights}")
    result = 0.0
    for i in range(len(weights)-1):
        result += values[i] * weights[i]
    result += 1 * weights[-1]
    #print(f"\ncalcFoward:{result}")
    return result

def calcCost(value, ideal):
    result = (ideal - value)**2/2
    #print(f"Cost:{result}")
    return result

def calcdETotalGo(value, ideal):
    result = -(ideal - value)
    #print(f"dETotal:{result}")
    return result

In [98]:
def relu(value):
    if(value < 0):
        value = 0
    print(f"\relu:{value}")
    return value

def drelu(value):
    if(value > 0):
        return value
    else:
        return 0

def linear(value):
    return value

def dlinear(value):
    return 1

def sigmoid(value):
    try:
        value = (1/(1 + np.exp(-value)))
    except RuntimeWarning:
        value = np.finfo(np.float64).tiny
    return value

def dsigmoid(value):
    value = value * (1 - value)
    return value

In [99]:
class Layer:
    def __init__(self,layer_size, neuron_size, activation_function, dactivation_function):
        self.neurons = np.random.rand(layer_size, neuron_size+1)
        #matrix = np.array(weights)
        #self.neurons = matrix
        print(f"\nNEURONS:{self.neurons}")
        self.activation_function = activation_function
        self.dactivation_function = dactivation_function
    
    def updateWeights(self, update_values, alpha):
        #print(f"Udate Weights\nUpdate Values:{update_values}\nAlpha:{alpha}")
        for i in range(len(self.neurons)):
            #print(f"\nNeuron {i}")
            for j in range(len(self.neurons[i])):
                #print(f"\nWeight {j}")
                #print(f"\nValue:{update_values[i][j]}")
                #print(f"Equation\n{self.neurons[i][j]} - {alpha} * {update_values[i][j]}")
                self.neurons[i][j] = self.neurons[i][j] - alpha * update_values[i][j]


    def foward(self,values):
        layer_a = []
        layer_z = []
        #print(f"\nValues:{values}")
        for i in range(len(self.neurons)):
            #print(f"\nNeuron:{i}")
            z = calcFoward(values,self.neurons[i])
            a = self.activation_function(z)
            #print(f"a:{a}\nz:{z}")
            layer_a.append(a)
            layer_z.append(z)
            
        #print(f"\nLayer A:{layer_a}\nLayer Z:{layer_z}")
        return np.array(layer_a),np.array(layer_z)
    
    def backpropagation(self, a0, a1, z0, ideals):
        #print(f"A0:{a0}\nA1:{a1}\nZ:{z0}\nIdeals:{ideals}")
        result = []
        aux = []
        for i in range(len(self.neurons)):
            neuron = []
            #print(f"\nNeuron:{i}")
            dEg = calcdETotalGo(a0[i],ideals[i])
            #print(f"dEg:{dEg}")
            dgouo = self.dactivation_function(a0[i])
            #print(f"dgouo:{dgouo}")
            #pesos
            for j in range(len(self.neurons[i])-1):
                #print(f"\nWeight:{j}")
                dgh = a1[j]
                #print(f"dgh:{dgh}")
                dEw = dEg * dgouo * dgh
                #print(f"dEw:{dEw}")
                neuron.append(dEw)
            #bias
            #print("\nBias:")
            dEw = dEg * dgouo
            #print(f"dEw:{dEw}")
            neuron.append(dEw)
            #print(f"Fix Result:{neuron}")
            result.append(np.array(neuron))
            aux.append(dEw)
            #print(f"aux {i}:\n{aux}")
        return np.array(result), np.array(aux)

    def backpropagationHidden(self, dEu, a0, a1, previous_layer):
        #print(f"\nA0:{a0}\nA1:{a1}\ndEu:{dEu}\nWeights:{self.neurons}\nPrevious Layer:{previous_layer}")
        result = []
        for i in range(len(self.neurons)):
            neuron = []
            #print(f"\nNeuron:{i}")
            dghuh = self.dactivation_function(a0[i])
            #print(f"dghuh:{dghuh}")
            true_dEg = []
            for j in range(len(previous_layer)):
                for k in range(len(dEu)):
                    #print(f"\nCalc dEg:{j}")
                    dEg = previous_layer[j][i] * dEu[k] 
                    #print(f"W:{previous_layer[j][i]}\ndEu:{dEu[j]}\ndEg:{dEg}")
                    true_dEg.append(dEg)
            true_dEg = sum(true_dEg)
            #print(f"\nTrue dEg:{true_dEg}")
            #pesos
            for k in range(len(self.neurons[i])-1):
                #print(f"\nWeight:{k}")
                dgh = a1[k]
                #print(f"dgh:{dgh}")
                dEw = true_dEg * dghuh * dgh
                #print(f"\ndEw:{dEw}")
                neuron.append(dEw)
            #bias
            #print(f"\nBias:")
            dEw = true_dEg * dghuh
            #print(f"\ndEw:{dEw}")
            neuron.append(dEw)
            result.append(np.array(neuron))


        return np.array(result)


    def relat(self):
        print(self.neurons)

In [100]:
class Network:
    def __init__(self):
        self.layers = []

    def appendLayer(self, layer_size, neuron_size, activation_function, dactivation_function):
        self.layers.append(Layer(layer_size, neuron_size, activation_function, dactivation_function))

    def foward(self,values):
        #print(f"\nFirst_Values:{values}")
        layer_a = [values]
        layer_z = []
        for i in range(len(self.layers)):
            #print(f"\nFoward:{i}")
            values,z = self.layers[i].foward(values)
            layer_a.append(values)
            layer_z.append(z)
        return layer_a,layer_z
                
    def backpropagation(self, all_a, all_z, ideals):
        #print(f"\nBack First")
        #printM3(all_a,"All A")
        #printM3(all_z,"All Z")
        #print(f"\nIdeals:{ideals}")
        result = []
        next = []
        for i in range(len(self.layers)):
            if i == 0:
                #print(f"\nBack")
                aux, next = self.layers[-(i+1)].backpropagation(all_a[-(i+1)], all_a[-(i+2)], all_z[-(i+1)], ideals)
                result.append(aux)
                #print(next)
            else:
                #print(f"\nBack Hidden {i}")
                aux = self.layers[-(i+1)].backpropagationHidden(next, all_a[-(i+1)], all_a[-(i+2)], self.layers[-i].neurons)
                result.append(aux)
        return result
        
    def calcError(self, output, ideals):
        result = 0
        #print(f"Ideals:{ideals}\nOut:{output}")
        for i in range(len(output)):
            result = result + calcCost(output[i], ideals[i])
        #print(f"Total cost:{result}")
        return result
    
    def updateWeights(self, update_values, alpha):
        for i in range(len(self.layers)):
            self.layers[-(i+1)].updateWeights(update_values[i],alpha)

    def train(self, input, ideals, alpha, epoch):
        print("Training")
        acumulated = 0
        for i in range(epoch):
            error = []
            for j in range(len(input)):
                all_a, all_z = self.foward(input[j])
                #print(f"\nAll Activations:\n{all_a}")
                if(i <= 2 or i >= 998):
                    error.append(self.calcError(all_a[-1],ideals[j]))
                fix = self.backpropagation(all_a, all_z, ideals[j])
                #print(f"fix:{fix}")
                if type(acumulated) is int:
                    acumulated = fix
                else:
                    for k in range(len(acumulated)):
                        for l in range(len(acumulated[k])):
                            acumulated[k][l] = acumulated[k][l] + fix[k][l]
                #print(f"\nacc {j}:\n{acumulated}")
            if(i <= 2 or i >= epoch-2):
                error = sum(error)
                error = error / len(input)
                print(f"Total Error Epoch {i}: {error}")
            #print(f"Acumulated sum:{acumulated}")
            for k in range(len(acumulated)):
                for l in range(len(acumulated[k])):
                    acumulated[k][l] = acumulated[k][l] / len(input)
            #print(f"Acumulated div:{acumulated}")
            self.updateWeights(acumulated,alpha)

    def test(self, input, ideals):
        print("Testing")
        error = []
        for j in range(len(input)):
            all_a, all_z = self.foward(input[j])
            error.append(self.calcError(all_a[-1],ideals[j]))
        error = sum(error)
        error = error / len(input)
        print(f"Total Error:{error}")
    
    def relat(self):
        for i in range(len(self.layers)):
            print(f"\nLayer{i}:")
            self.layers[i].relat()


In [101]:
network = Network()
network.appendLayer(8,4,sigmoid,dsigmoid)
network.appendLayer(4,8,sigmoid,dsigmoid)
network.appendLayer(1,4,sigmoid,dsigmoid)


NEURONS:[[0.53505281 0.48067163 0.01322175 0.89639771 0.88677219]
 [0.03482259 0.92567307 0.53967346 0.69864528 0.69220854]
 [0.19529239 0.00302992 0.44629357 0.5240007  0.79206165]
 [0.95705507 0.88678711 0.36716265 0.43427406 0.69204985]
 [0.00763542 0.05667083 0.73754472 0.9619466  0.51904677]
 [0.45343916 0.06253496 0.62399692 0.94321288 0.39924876]
 [0.49617239 0.98929866 0.41295446 0.45439981 0.17472523]
 [0.09275761 0.65247683 0.01826401 0.15739375 0.82830725]]

NEURONS:[[0.44537192 0.21518127 0.41157327 0.34226803 0.31596793 0.94241969
  0.98593782 0.6898981  0.45942567]
 [0.7302051  0.65108849 0.11908495 0.13888099 0.62663989 0.52569359
  0.71893059 0.63195558 0.14067143]
 [0.04703326 0.93823256 0.41383944 0.75651595 0.4575987  0.91665408
  0.36531643 0.48181913 0.54214596]
 [0.4750006  0.24654197 0.88258276 0.44431443 0.16435661 0.26430209
  0.13308391 0.25551694 0.38146345]]

NEURONS:[[0.52883315 0.64622763 0.08830572 0.17138982 0.9301475 ]]


In [102]:
def add_offset(X):
    n_samples = X.shape[0]
    n_features = X.shape[1]

    X_with_offset = np.zeros((n_samples, n_features + 1))

    for i in range(n_samples):
        X_with_offset[i, 0] = 1.0
    
    for i in range(n_samples):
        for j in range(n_features):
            X_with_offset[i, j + 1] = X[i, j]

    X = X_with_offset

    return X

# normaliza as colunas de características (0 a 1)
def normalize_x(X):
    X = (X - X.mean()) / X.std()
    return X

# function to prepare X (normalize + add a offset)
def prepare_X(X):
    return add_offset(normalize_x(X))

# map every possible result in a class number
def prepare_Y(y):
    mapping = {
        "Iris-setosa": 0,
        "Iris-versicolor": 1,
        "Iris-virginica": 2,
    }

    return y.map(mapping)

# Definição do dataframe iris Flowers
df = pd.read_csv('IRIS.csv')


X = df.drop(columns=['species']).to_numpy()

Y = prepare_Y(df['species']).to_numpy()

# Divisão entre treino e teste
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

clean_input = []
for i in range(len(X)):
    clean_input.append(np.array(X[i]))

Y = Y.tolist()

clean_all_ideals = []
for i in range(len(Y)):
    aux = [Y[i]]
    clean_all_ideals.append(aux)

#print(clean_all_ideals)

network.relat()

network.train(clean_input, clean_all_ideals, 0.5, 10000)

network.test(clean_input,clean_all_ideals)

network.relat()


Layer0:
[[0.53505281 0.48067163 0.01322175 0.89639771 0.88677219]
 [0.03482259 0.92567307 0.53967346 0.69864528 0.69220854]
 [0.19529239 0.00302992 0.44629357 0.5240007  0.79206165]
 [0.95705507 0.88678711 0.36716265 0.43427406 0.69204985]
 [0.00763542 0.05667083 0.73754472 0.9619466  0.51904677]
 [0.45343916 0.06253496 0.62399692 0.94321288 0.39924876]
 [0.49617239 0.98929866 0.41295446 0.45439981 0.17472523]
 [0.09275761 0.65247683 0.01826401 0.15739375 0.82830725]]

Layer1:
[[0.44537192 0.21518127 0.41157327 0.34226803 0.31596793 0.94241969
  0.98593782 0.6898981  0.45942567]
 [0.7302051  0.65108849 0.11908495 0.13888099 0.62663989 0.52569359
  0.71893059 0.63195558 0.14067143]
 [0.04703326 0.93823256 0.41383944 0.75651595 0.4575987  0.91665408
  0.36531643 0.48181913 0.54214596]
 [0.4750006  0.24654197 0.88258276 0.44431443 0.16435661 0.26430209
  0.13308391 0.25551694 0.38146345]]

Layer2:
[[0.52883315 0.64622763 0.08830572 0.17138982 0.9301475 ]]
Training
Total Error Epoch 0: 0.

network2 = Network2()
network2.appendLayer(2,2,sigmoid,dsigmoid,[[0.15, 0.2, 0.35], [0.25, 0.3, 0.35]])
network2.appendLayer(2,2,sigmoid,dsigmoid,[[0.4, 0.45, 0.6], [0.5, 0.55, 0.6]])

input = [[1,1],[0.05,0.1],[1,0],[0,0]]
all_ideals = [[0,1],[0.01,0.99],[1,0],[0,1]]

clean_input = []
for i in range(len(input)):
    clean_input.append(np.array(input[i]))

clean_all_ideals = np.array(all_ideals)

network2.train([clean_input[1]], [clean_all_ideals[1]], 0.5, 10000)